v2/internal/controllers/documentdb_sql_crud_v20231115_test.go

```diff
@@ -0,0 +1,440 @@
+				Id: &name,
+				ConflictResolutionPolicy: &documentdb.ConflictResolutionPolicy{
+					Mode: &lastWriterWins,
+				},
+				DefaultTtl: to.Ptr(200),
+				IndexingPolicy: &documentdb.IndexingPolicy{
+					IndexingMode: &consistent,
+					IncludedPaths: []documentdb.IncludedPath{{
+						Path: to.Ptr("/*"),
+					}},
+					ExcludedPaths: []documentdb.ExcludedPath{{
+						Path: to.Ptr("/myPathToNotIndex/*"),
+					}},
+				},
+				PartitionKey: &documentdb.ContainerPartitionKey{
+					Kind:  &hash,
+					Paths: []string{"/myPartitionKey"},
+				},
+			},
+		},
+	}
+
+	tc.CreateResourceAndWait(&container)
+	defer func() {
+		tc.LogSubsectionf("Cleaning up container %q", name)
+		tc.DeleteResourceAndWait(&container)
+	}()
+
+	tc.RunParallelSubtests(
+		testcommon.Subtest{
+			Name: "CosmosDB SQL Trigger CRUD",
+			Test: func(tc *testcommon.KubePerTestContext) {
+				CosmosDB_SQL_Trigger_v20231115_CRUD(tc, &container)
+			},
+		},
+		testcommon.Subtest{
+			Name: "CosmosDB SQL Stored Procedure CRUD",
+			Test: func(tc *testcommon.KubePerTestContext) {
+				CosmosDB_SQL_StoredProcedure_v20231115_CRUD(tc, &container)
+			},
+		},
+		testcommon.Subtest{
+			Name: "CosmosDB SQL User-defined Function CRUD",
+			Test: func(tc *testcommon.KubePerTestContext) {
+				CosmosDB_SQL_UserDefinedFunction_v20231115_CRUD(tc, &container)
+			},
+		},
+		testcommon.Subtest{
+			Name: "CosmosDB SQL Container ThroughputSettings CRUD",
+			Test: func(tc *testcommon.KubePerTestContext) {
+				CosmosDB_SQL_Database_Container_ThroughputSettings_v20231115_CRUD(tc, &container)
+			},
+		})
+
+	tc.LogSubsectionf("Updating the default TTL on container %q", name)
+	old := container.DeepCopy()
+	container.Spec.Resource.DefaultTtl = to.Ptr(400)
+	tc.PatchResourceAndWait(old, &container)
+	tc.Expect(container.Status.Resource).ToNot(BeNil())
+	tc.Expect(container.Status.Resource.DefaultTtl).ToNot(BeNil())
+	tc.Expect(*container.Status.Resource.DefaultTtl).To(Equal(400))
+}
+
+func CosmosDB_SQL_Trigger_v20231115_CRUD(tc *testcommon.KubePerTestContext, container client.Object) {
+	tc.LogSectionf("Creating SQL Trigger")
+
+	// Declare a trigger
+	name := tc.Namer.GenerateName("trigger")
+	pre := documentdb.SqlTriggerResource_TriggerType_Pre
+	create := documentdb.SqlTriggerResource_TriggerOperation_Create
+	trigger := documentdb.SqlDatabaseContainerTrigger{
+		ObjectMeta: tc.MakeObjectMetaWithName(name),
+		Spec: documentdb.DatabaseAccounts_SqlDatabases_Containers_Trigger_Spec{
+			Location: to.Ptr("australiaeast"), // Capacity constraints // tc.AzureRegion
+			Owner:    testcommon.AsOwner(container),
+			Resource: &documentdb.SqlTriggerResource{
+				Id:               &name,
+				TriggerType:      &pre,
+				TriggerOperation: &create,
+				Body:             to.Ptr(triggerBody_v20231115),
+			},
+		},
+	}
+
+	tc.CreateResourceAndWait(&trigger)
+	defer func() {
+		tc.LogSubsectionf("Cleaning up trigger %q", name)
+		tc.DeleteResourceAndWait(&trigger)
```

Comment:
FWIW you don't actually need to do this `defer` trickery, because we track all the resources that are created and they get unrolled automatically by the test framework at the end.

I mean this doesn't hurt but it's not doing a ton either (unless there's some specific delete ordering issue you're trying to solve?)

Reply:
Replicating code from the earlier version - happy to tidy this up